### PR TITLE
Support for piecewise (convex) costs in ACOPF

### PR DESCRIPTION
**Pull request by michaelbynum**
Support for piecewise (convex) costs in ACOPF

### CODE REVIEW

File: egret/model_library/transmission/gen.py [hunk 1]
```diff
@@ -28,6 +28,65 @@ def declare_var_qg(model, index_set, **kwargs):
...
+            pt2, cost2 = cost_dict['values'][ndx + 1]
+            slope = (cost2 - cost1) / (pt2 - pt1)
+            if last_slope is not None:
+                if slope < last_slope:
+                    all_slopes = [(cost_dict['values'][i + 1][1] - cost_dict['values'][i][1]) / (cost_dict['values'][i + 1][0] - cost_dict['values'][i][0]) for i in range(len(cost_dict['values']) - 1)]
+                    raise ValueError(f'Piecewise costs must be convex; generator: {gen_name}; slopes: {all_slopes}')
+            intercept = cost2 - slope * pt2
+            pw_cost_set.add((gen_name, ndx))
+            indexed_pw_cost_con[gen_name, ndx] = cost_var >= slope * gen_var + intercept
+            last_slope = slope
```
Comment (bknueven) [inline egret/model_library/transmission/gen.py:59]: At this point we could bound and initialize `cost_var`.

File: egret/models/acopf.py [hunk 2]
```diff
@@ -228,18 +228,31 @@ def _create_base_power_ac_model(model_data, include_feasibility_slack=False):
...
+        q_cost = dict()
+    for gen_name in gen_attrs['names']:
+        if p_cost[gen_name]['cost_curve_type'] == 'piecewise' or (gen_name in q_cost and q_cost[gen_name]['cost_curve_type'] == 'piecewise'):
+            pw_cost_gens.append(gen_name)
+
+    if len(pw_cost_gens) > 0:
+        libgen.declare_var_pg_cost(model=model, index_set=pw_cost_gens)
+        libgen.declare_var_qg_cost(model=model, index_set=pw_cost_gens)
+        libgen.declare_piecewise_cost_cons(model=model, index_set=pw_cost_gens, p_costs=p_cost, q_costs=q_cost)
+
```
Comment (bknueven) [inline egret/models/acopf.py:246]: It would be nice if it were easy to insert this logic into the other OPF models.

**Review by bknueven — APPROVED** [review]
I have lots of comments, mostly because I've implemented piecewise cost curves a few times; feel free to consider them completely optional.

The unit commitment model spends quite bit of effort validating and potentially simplifying the data passed in, e.g., ignoring segments outside of [p_min, p_max], combining close segments, etc. Unfortunately some of these boring and tedious things can make a performance difference; it's also nice just to have some of this preprocessing in Egret itself if the data is messy.

I guess there's also the question of if this is *the* PWL cost formulation for the ACOPF model, see: https://arxiv.org/pdf/2005.14087.pdf. This also discusses a bit of the preprocessing issues.

**Comment by michaelbynum** [comment]
@bknueven Thanks for all of the feedback. I think this is a greatly improved PR now. Let me know what you think.

**Comment by bknueven** [comment]
As it is, the validation code does not handle if the p_max <= the first piecewise point or the p_min >= the last piecewise point. We should either address this in this PR or open and issue.

**Pull request merged**